### PR TITLE
#11691: Move tensor module to ttnn

### PR DESCRIPTION
**Pull request by KalaivaniMCW**
### Ticket
Link to Github Issue #11691

### Problem description
Move tensor module to ttnn

### What's changed
- Move tensor module bindings from ttnn/deprecated to ttnn
- all references to ttnn.experimental.tensor.* no longer exist
- moved convert_conv_weight_tensor* apis from tensor module to convolution (conv2d)
- removed duplicated bindings of allocate_tensor_on_device, write_tensor 
- make Core* classes available in tt::tt_metal namespace

### Checklist
- [ ] Post commit CI https://github.com/tenstorrent/tt-metal/actions/runs/10670076608
https://github.com/tenstorrent/tt-metal/actions/runs/10809863242
- [ ] Demo tests https://github.com/tenstorrent/tt-metal/actions/runs/10670082858
as main - https://github.com/tenstorrent/tt-metal/actions/runs/10775974676
- [ ] t3k demos https://github.com/tenstorrent/tt-metal/actions/runs/10670095656
https://github.com/tenstorrent/tt-metal/actions/runs/10775981583
- [ ] t3k frequent https://github.com/tenstorrent/tt-metal/actions/runs/10670080283
https://github.com/tenstorrent/tt-metal/actions/runs/10775970710
- [ ] t3k unit https://github.com/tenstorrent/tt-metal/actions/runs/10670091567
https://github.com/tenstorrent/tt-metal/actions/runs/10775991646
- [ ] t3k model perf https://github.com/tenstorrent/tt-metal/actions/runs/10670086480
- [ ] nightly FD https://github.com/tenstorrent/tt-metal/actions/runs/10670088963
https://github.com/tenstorrent/tt-metal/actions/runs/10775985909
- [ ] device perf https://github.com/tenstorrent/tt-metal/actions/runs/10670084135
https://github.com/tenstorrent/tt-metal/actions/runs/10775977596
- [ ] pc model
- [ ] model perf https://github.com/tenstorrent/tt-metal/actions/runs/10670078386
https://github.com/tenstorrent/tt-metal/actions/runs/10777550272
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/sweep_framework/sweeps/matmul/short/matmul_user_program_config_mcast_1d.py [hunk 1]
```diff
@@ -103,7 +103,7 @@ class TensorMemoryConfigs(enum.Enum):
                 memory_layout=ttnn.TensorMemoryLayout.WIDTH_SHARDED,
                 buffer_type=ttnn.BufferType.L1,
                 shard_spec=ttnn.ShardSpec(
-                    ttnn.experimental.tensor.num_cores_to_core_range_set(35, core_grid, row_wise=True),
+                    ttnn._ttnn.tensor.num_cores_to_core_range_set(35, core_grid, row_wise=True),
                     (64, IN0_INNER_DIM_PER_CORE),
```
Comment (eyonland) [inline tests/sweep_framework/sweeps/matmul/short/matmul_user_program_config_mcast_1d.py:107]: I'm assuming you will change this to ttnn.num_cores_to_core_range_set

Comment (KalaivaniMCW) [reply tests/sweep_framework/sweeps/matmul/short/matmul_user_program_config_mcast_1d.py:107]: yes, it will be changed in @VirdhatchaniKN 's work split PR

File: tests/tt_eager/python_api_testing/conv/conv_unit_test_utils.py [hunk 2]
```diff
@@ -50,7 +50,7 @@ def create_conv_weight_tensor(torch_tensor, K, C, R, S, in1_block_h, in1_block_w
     B_ = ttnn.Tensor(torch.flatten(torch_tensor).tolist(), weights_shape, ttnn.bfloat16, ttnn.ROW_MAJOR_LAYOUT).pad(
         weights_channels_padded_shape, (0, 0, 0, 0), 0.0
     )
-    B_tiled_host = ttnn.experimental.tensor.convert_conv_weight_tensor_to_tiled_layout(B_, in1_block_h, in1_block_w)
+    B_tiled_host = ttnn.tensor.convert_conv_weight_tensor_to_tiled_layout(B_, in1_block_h, in1_block_w)
     return B_tiled_host
```
Comment (eyonland) [inline tests/tt_eager/python_api_testing/conv/conv_unit_test_utils.py:54]: @mywoodstock , do you have an opinion on this?
Should this be:
ttnn.conv.convert_conv_weight_tensor_to_special_padding_tiled_layout
or
ttnn.convert_conv_weight_tensor_to_special_padding_tiled_layout

Comment (mywoodstock) [reply tests/tt_eager/python_api_testing/conv/conv_unit_test_utils.py:54]: I don't have a preference, but im thinking `conv.convert...` makes more sense since its conv specific helper function.

Comment (KalaivaniMCW) [reply tests/tt_eager/python_api_testing/conv/conv_unit_test_utils.py:54]: I have moved them from tensor module to conv2d module @mywoodstock @eyonland

**Review by Aswinmcw — APPROVED** [review]
LGTM

File: tt_metal/common/core_coord.h [hunk 3]
```diff
@@ -20,6 +20,14 @@ using std::pair;
 
 using CoreCoord = tt_xy_pair;
 
+// Adding to tt::tt_metal namespace as we transition to moving this out of global namespace eventually.
+namespace tt::tt_metal {
+   using ::CoreCoord;
+}
+
+namespace ttnn::tensor {
```
Comment (ayerofieiev-tt) [inline tt_metal/common/core_coord.h:28]: tt_metal is a lower level library and should know nothing about ttnn

File: ttnn/cpp/ttnn/tensor/tensor.hpp [hunk 4]
```diff
@@ -282,7 +282,7 @@ struct Tensor {
     //                                  Low Level APIs
     // ======================================================================================
     Tensor reshape(int N, int C, int H, int W) const;
-    Tensor reshape(const Shape &new_shape) const;
+    Tensor reshape(const tt::tt_metal::Shape &new_shape) const;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/tensor.hpp:285]: ❤️

File: ttnn/ttnn/operations/core.py [hunk 5]
```diff
@@ -580,7 +579,7 @@ def from_torch_and_dump(
             )
             pathlib.Path(cache_file_name).parent.mkdir(parents=True, exist_ok=True)
             distributed_config = mesh_mapper.config() if mesh_mapper else dict()
-            ttnn.dump_tensor(cache_file_name, tensor, distributed_config)
```
Comment (ayerofieiev-tt) [inline ttnn/ttnn/operations/core.py:583]: Why this change happened? Is it intentional?

Comment (KalaivaniMCW) [reply ttnn/ttnn/operations/core.py:583]: previously ttnn.dump_tensor api  just called tt_lib.tensor.dump_tensor api
now that the binding for dump_tensor is moved to a ttnn submodule, makes sense to use it.

**Review by mywoodstock — APPROVED** [review]

File: ttnn/cpp/pybind11/tensor.hpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+namespace ttnn::tensor {
+
+namespace detail{
+        void pytensor_module_types(pybind11::module & m_tensor);
+        void pytensor_module(pybind11::module & m_tensor);
+}
+
+    void py_module_types(py::module& module);
+    void py_module(py::module& module);
+
```
[the remark this answers is not in the document]
Comment (KalaivaniMCW) [reply ttnn/cpp/pybind11/tensor.hpp:20]: @ayerofieiev-tt are you referring to the py_module (tensor.cpp) and pytensor_module (pytensor.cpp) module being separate ? It was kept separate earlier I am guessing for some particular reason (pytensor contains a lot of convert functions that are used inside the bindings, and it contains necessary documentation for Tensor and related functions)
As for separating 
@yan-zaretskiy suggested we keep the types( enums and classes) and function under separate binding calls and types precede function. https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/pybind11/__init__.cpp

File: tt_metal/common/core_coord.h (hunk 3 above)
Comment (KalaivaniMCW) [reply tt_metal/common/core_coord.h:28]: @eyonland 's suggestion ( a temp one) until CoreCoord and other necessary classes are moved out of global namespace. are you suggesting we remove namespace ttnn::tensor and keep namespace tt::tt_metal for now ?

File: ttnn/cpp/pybind11/tensor.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (KalaivaniMCW) [reply ttnn/cpp/pybind11/tensor.hpp:20]: I did try to have it as only 2 functions - 1 for types and 1 for function call, but then again, it looks better to have pytensor.cpp separately as it is dedicated to the Tensor page in our TTNN documentation

File: tt_metal/common/core_coord.h (hunk 3 above)
Comment (eyonland) [reply tt_metal/common/core_coord.h:28]: @KalaivaniMCW  we for sure don't want to do this here.

My suggestion is to move this namespace declaration that defines ttnn::tensor::CoreCoord to ttnn/cpp/ttnn/tensor/types.hpp

**Review by pavlejosipovic — APPROVED** [review]

File: tt_metal/host_api.hpp [hunk 7]
```diff
@@ -26,6 +26,17 @@
 class CoreRange;
 class CoreRangeSet;
 
+// Adding to tt::tt_metal namespace as we transition to moving this out of global namespace eventually.
+namespace tt::tt_metal {
+   using ::CoreRange;
+   using ::CoreRangeSet;
+}
```
Comment (tt-aho) [inline tt_metal/host_api.hpp:33]: I think this should go inside `core_coord.h` together with what you did for `CoreCoord` instead of putting it here.

**Review by tt-aho — CHANGES_REQUESTED** [review]

File: tt_metal/host_api.hpp [hunk 8]
```diff
@@ -26,6 +26,17 @@
...
+
+namespace ttnn::tensor {
+   using tt::tt_metal::CoreRange;
+   using tt::tt_metal::CoreRangeSet;
+}
```
[the remark this answers is not in the document]
Comment (KalaivaniMCW) [reply tt_metal/host_api.hpp]: removed

File: tt_metal/host_api.hpp (hunk 7 above)
Comment (KalaivaniMCW) [reply tt_metal/host_api.hpp]: Done. `CoreRange, CoreRangeSet, CoreCoord`  are all placed in  `core_coord.h`.

**Review by tt-aho — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

**Review by sjameelTT — APPROVED** [review]

**Review by jdesousa-TT — APPROVED** [review]

**Comment by KalaivaniMCW** [comment]
Hi @razorback3, @dongjin-na, 
Could you please review this PR ?

**Review by eyonland — APPROVED** [review]

File: ttnn/cpp/pybind11/tensor.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (ayerofieiev-tt) [reply ttnn/cpp/pybind11/tensor.hpp:20]: @KalaivaniMCW 2 things:
* Code indentation is wrong here.
* Not clear why detail namespace is in the header. I assume that functions declared in `details` namespace are only used in a corresponding pybind11/tensor.cpp, so declaration can be removed from .hpp
* It is weird from a caller standpoint to know the difference between functions `py_module` and `py_module_types` which are exposed in tensor.hpp. I recommend having a single function in a header to bind tensor (and whatever else it provides). In .cpp it can call to multiple functions declared/defined in .cpp, not exposed in .hpp

File: ttnn/ttnn/operations/core.py (hunk 5 above)
Comment (ayerofieiev-tt) [reply ttnn/ttnn/operations/core.py:599]: But can we expose `dump_tensor` from `ttnn._ttnn.tensor` into `ttnn`?

**Review by ayerofieiev-tt — APPROVED** [review]
Approving to not block when the feedback is resolved
